File: typescript/src/x402-decode.ts
/**
 * Decode and validate the X-PAYMENT header returned by the Canopy backend.
 *
 * The header is base64(JSON({ x402Version, scheme, network, payload: {
 * signature, authorization: { from, to, value, validAfter, validBefore,
 * nonce } } })) per the x402 spec. We decode it client-side as a
 * defense-in-depth check before retrying the resource server: a bug or
 * misuse in the backend that returns a header for the wrong offer (or an
 * expired one) is caught here rather than silently approving an unintended
 * payment.
 */

interface XPaymentEnvelope {
  x402Version: number;
  scheme: string;
  network: string;
  payload: {
    signature: string;
    authorization: {
      from: string;
      to: string;
      value: string;
      validAfter: string;
      validBefore: string;
      nonce: string;
    };
  };
}

function fromBase64Utf8(s: string): string {
  if (typeof Buffer !== "undefined") return Buffer.from(s, "base64").toString("utf8");
  // Browser fallback
  const bin = atob(s);
  const bytes = new Uint8Array(bin.length);
  for (let i = 0; i < bin.length; i++) bytes[i] = bin.charCodeAt(i);
  return new TextDecoder("utf-8").decode(bytes);
}

export function decodeXPaymentHeader(header: string): XPaymentEnvelope | null {
  try {
    const parsed = JSON.parse(fromBase64Utf8(header));
    if (
      typeof parsed?.payload?.authorization?.to !== "string" ||
      typeof parsed?.payload?.authorization?.validBefore !== "string"
    ) {
      return null;
    }
    return parsed as XPaymentEnvelope;
  } catch {
    return null;
  }
}

export function verifyXPaymentMatchesOffer(
  header: string,
  offer: { payTo: string },
): { ok: true } | { ok: false; reason: string } {
  const envelope = decodeXPaymentHeader(header);
  if (!envelope) return { ok: false, reason: "X-PAYMENT header is not a valid envelope" };

  const auth = envelope.payload.authorization;
  if (auth.to.toLowerCase() !== offer.payTo.toLowerCase()) {
    return { ok: false, reason: "X-PAYMENT recipient does not match the 402 offer" };
  }

  const validBefore = Number(auth.validBefore);
  if (!Number.isFinite(validBefore) || validBefore <= Math.floor(Date.now() / 1000)) {
    return { ok: false, reason: "X-PAYMENT is expired" };
  }

  return { ok: true };
}

File: python/canopy_ai/x402_decode.py
"""
Decode and validate the X-PAYMENT header returned by the Canopy backend.

The header is base64(JSON({ x402Version, scheme, network, payload: {
signature, authorization: { from, to, value, validAfter, validBefore,
nonce } } })) per the x402 spec. We decode it client-side as a defense-
in-depth check before retrying the resource server: a bug or misuse in
the backend that returns a header for the wrong offer (or an expired
one) is caught here rather than silently approving an unintended
payment.
"""

from __future__ import annotations

import base64
import json
import time
from typing import Any


def decode_x_payment_header(header: str) -> dict[str, Any] | None:
    try:
        decoded = base64.b64decode(header.encode("ascii"), validate=False)
        parsed = json.loads(decoded.decode("utf-8"))
    except (ValueError, json.JSONDecodeError):
        return None

    if not isinstance(parsed, dict):
        return None
    payload = parsed.get("payload")
    if not isinstance(payload, dict):
        return None
    auth = payload.get("authorization")
    if not isinstance(auth, dict):
        return None
    if not isinstance(auth.get("to"), str) or not isinstance(auth.get("validBefore"), str):
        return None
    return parsed


def verify_x_payment_matches_offer(header: str, offer: dict[str, Any]) -> tuple[bool, str | None]:
    envelope = decode_x_payment_header(header)
    if envelope is None:
        return False, "X-PAYMENT header is not a valid envelope"

    auth = envelope["payload"]["authorization"]
    pay_to = offer.get("payTo")
    if not isinstance(pay_to, str):
        return False, "Offer is missing payTo"

    if auth["to"].lower() != pay_to.lower():
        return False, "X-PAYMENT recipient does not match the 402 offer"

    try:
        valid_before = int(auth["validBefore"])
    except (TypeError, ValueError):
        return False, "X-PAYMENT validBefore is malformed"

    if valid_before <= int(time.time()):
        return False, "X-PAYMENT is expired"

    return True, None
